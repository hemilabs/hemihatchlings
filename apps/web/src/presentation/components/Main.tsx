import React from 'react'

function Main(): JSX.Element {
  return (
    <h1 className='text-2xl'>Home Page</h1>
  )
}

export default Main
