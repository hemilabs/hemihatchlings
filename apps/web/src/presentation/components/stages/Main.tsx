import { useContext, useEffect, useState } from 'react'
import { servicesContext } from '../../context'
import ProgressBar from './ProgressBar'
import OptInButton from './OptInButton'
import Eggs from './Eggs'
import Hatchling from './Hatchling'
import StatusMessage from './StatusMessage'
import { ElementEnum, Stage, StageEnum } from '@hemihatchlings/shared'

function Main(): JSX.Element {
  const [stage, setStage] = useState<StageEnum>(StageEnum.Egg)
  const [address, setAddress] = useState('')
  const [transactionHash, setTransactionHash] = useState('')
  const [element, setElement] = useState<ElementEnum>(ElementEnum.Fire)

  const {
    connectWalletService,
    mintNFTService,
    getHatchlingStageService
  } = useContext(servicesContext)

  const connectWallet = async () => {
    const response = await connectWalletService.execute()

    setAddress(response.address)
  }

  const mintNFT = async () => {
    const response = await mintNFTService.execute()

    setTransactionHash(response.transactionHash)
  }

  const hatchEgg = async (element: ElementEnum) => {
    await connectWallet()
    await mintNFT()

    setElement(element)
    setStage(StageEnum.Baby)
  }

  const fetchHatchlingStage = async () => {
    const response = await getHatchlingStageService.execute({ transactionHash })
    const stageValue = response?.stage as keyof typeof StageEnum
    
    setStage(StageEnum[stageValue])
  }

  useEffect(() => {
    if (stage !== StageEnum.Egg) {
      fetchHatchlingStage()
    }
  }, [])

  return (
  <>
    <div className={stage === StageEnum.Egg ? '' : 'hidden'}>
      <Eggs hatchEgg={hatchEgg}/>
    </div>

    <div className={stage != StageEnum.Egg ? '' : 'hidden'}>
      <Hatchling element={element} stage={stage}/>

      <div className="mt-4 sm:mt-8 lg:mt-12">
        <OptInButton />
      </div>
      <div className="mt-4 sm:mt-8 lg:mt-12">
        <ProgressBar percent={8}/>
        {address}
        <StatusMessage stage={stage} />
      </div>
    </div>
  </>
  )
}

export default Main
