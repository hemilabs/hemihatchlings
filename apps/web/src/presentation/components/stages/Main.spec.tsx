import { describe, it, expect} from 'vitest'
import { RenderResult, render } from '@testing-library/react'
import Main from './Main'
import {
  ServicesContext,
  servicesContext,
  defaultValue
} from '../../context'

const renderComponent =
  (context: ServicesContext = defaultValue): RenderResult => {
    return render(
      <servicesContext.Provider value={ context }>
        <Main />
      </servicesContext.Provider>
    )
  }

describe('src/presentation/stages/Main', () => {
  it('renders correctly', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })
})
