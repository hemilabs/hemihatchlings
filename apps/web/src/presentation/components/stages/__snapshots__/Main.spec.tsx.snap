// Vitest Snapshot v1

exports[`src/presentation/stages/Main > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="opacity-0
        transition-opacity ease-in-out delay-150 duration-300 
        fixed top-3 right-3  items-center w-full max-w-xs p-4 mb-4
        rounded-lg shadow text-white bg-red-800"
    id="toast-danger"
    role="alert"
  >
    <div
      class="flex flex-row items-center"
    >
      <div
        class="inline-flex items-center justify-center flex-shrink-0 w-8 h-8 rounded-lg bg-red-500 text-red-200"
      >
        <svg
          aria-hidden="true"
          class="w-5 h-5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5Zm3.707 11.793a1 1 0 1 1-1.414 1.414L10 11.414l-2.293 2.293a1 1 0 0 1-1.414-1.414L8.586 10 6.293 7.707a1 1 0 0 1 1.414-1.414L10 8.586l2.293-2.293a1 1 0 0 1 1.414 1.414L11.414 10l2.293 2.293Z"
          />
        </svg>
      </div>
      <div
        class="ms-3 text-sm font-normal"
      />
    </div>
  </div>
  <div
    class=""
  >
    <div
      class="flex flex-row justify-center gap-4 lg:gap-12"
    >
      <button
        class="disabled:opacity-50 lg:hidden"
        disabled=""
      >
        <img
          alt="Go Left"
          class="w-32 rotate-180"
          src="/src/assets/chevron.svg"
        />
      </button>
      <div
        class=""
      >
        <button>
          <div
            class="border border-white transition duration-500 hover:scale-110"
          >
            <img
              alt="Fire Egg"
              src="./testnet/fire-egg.png"
            />
          </div>
          <div
            class="font-monaco text-sm text-white mt-2 text-left sm:text-base md:text-lg md:mt-4"
          >
            01
          </div>
        </button>
      </div>
      <div
        class="hidden lg:block"
      >
        <button>
          <div
            class="border border-white transition duration-500 hover:scale-110"
          >
            <img
              alt="Water Egg"
              src="./testnet/water-egg.png"
            />
          </div>
          <div
            class="font-monaco text-sm text-white mt-2 text-left sm:text-base md:text-lg md:mt-4"
          >
            02
          </div>
        </button>
      </div>
      <div
        class="hidden lg:block"
      >
        <button>
          <div
            class="border border-white transition duration-500 hover:scale-110"
          >
            <img
              alt="Earth Egg"
              src="./testnet/earth-egg.png"
            />
          </div>
          <div
            class="font-monaco text-sm text-white mt-2 text-left sm:text-base md:text-lg md:mt-4"
          >
            03
          </div>
        </button>
      </div>
      <button
        class="disabled:opacity-50 lg:hidden"
      >
        <img
          alt="Go Right"
          class="w-32"
          src="/src/assets/chevron.svg"
        />
      </button>
    </div>
    <div
      class="flex flex-col items-center"
    >
      <h2
        class="font-monaco font-normal text-white text-lg mt-8 sm:text-xl md:text-2xl lg:text-3xl lg:mt-12 xl:text-4xl"
      >
        FREE MINT + GAS FEES
      </h2>
    </div>
  </div>
  <div
    class="hidden"
  >
    <div
      class="flex flex-row justify-center"
    >
      <div
        class="flex flex-col"
      >
        <div
          class="
          border w-96 items-center"
        >
          <img
            alt="Fire Baby"
            class=""
            src="./testnet/fire-baby.png"
          />
          <img
            alt="Water Baby"
            class="hidden"
            src="./testnet/water-baby.png"
          />
          <img
            alt="Earth Baby"
            class="hidden"
            src="./testnet/earth-baby.png"
          />
          <img
            alt="Air Baby"
            class="hidden"
            src="./testnet/air-baby.png"
          />
        </div>
        <div
          class="font-monaco text-sm text-white mt-2 text-center sm:text-base md:text-lg md:mt-4"
        >
          âœ… HATCHING STAGE 1
        </div>
      </div>
      <div
        class="flex flex-col"
      >
        <div
          class="blur
          border w-96 items-center"
        >
          <img
            alt="Fire Adult"
            class=""
            src="./testnet/fire-adult.png"
          />
          <img
            alt="Water Adult"
            class="hidden"
            src="./testnet/water-adult.png"
          />
          <img
            alt="Earth Adult"
            class="hidden"
            src="./testnet/earth-adult.png"
          />
          <img
            alt="Air Adult"
            class="hidden"
            src="./testnet/air-adult.png"
          />
        </div>
        <div
          class="font-monaco text-sm text-white mt-2 text-center sm:text-base md:text-lg md:mt-4"
        >
          ðŸ”„ HATCHING STAGE 2
        </div>
      </div>
      <img
        alt="Lock Icon"
        class="
            absolute m-auto top-0 bottom-0 left-0 right-0"
        src="/src/assets/lock-icon.png"
      />
    </div>
    <div
      class="mt-4 sm:mt-8 lg:mt-12"
    >
      <div
        class="
        font-monaco text-sm text-white mt-2 text-center uppercase
        sm:text-base md:text-lg md:mt-4"
      >
        Approaching BITCOIN Finality: EST. 90 MIN WAIT.
      </div>
      <div
        class="hidden
        font-monaco text-sm text-white mt-2 text-center uppercase
        sm:text-base md:text-lg md:mt-4"
      >
        ethereum & bitcoin finality achieved
      </div>
      <div
        class="font-monaco text-sm text-white mt-2 text-center sm:text-base md:text-lg md:mt-4"
      >
        <div>
          Wallet: 
        </div>
        <div>
          Transaction: 
        </div>
      </div>
      <div
        class="text-white text-center underline"
      >
        <a
          href="https://testnet.explorer.hemi.network/tx/"
          target="_blank"
        >
          View on block explorer
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
